File: Day23/LongValidPar.java
class LongValidPar {
    public int longestValidParentheses(String s) {
        java.util.Stack<Integer> st = new java.util.Stack<>();
        st.push(-1);
        int maxLen = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                st.push(i);
            } else {
                st.pop();
                if (st.isEmpty()) {
                    st.push(i);
                } else {
                    maxLen = Math.max(maxLen, i - st.peek());
                }
            }
        }
        return maxLen;
    }

    public static void main(String[] args) {
        LongValidPar sol = new LongValidPar();
        String s = "(()";
        System.out.println(sol.longestValidParentheses(s)); 
    }
}

File: Day23/LongValidPar.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        st = [-1]
        max_len = 0

        for i in range(len(s)):
            if s[i] == '(':
                st.append(i)
            else:
                st.pop()
                if not st:
                    st.append(i)
                else:
                    max_len = max(max_len, i - st[-1])

        return max_len

if __name__ == '__main__':
    s = "(()"
    sol = Solution()
    print(sol.longestValidParentheses(s))  
